File: Source/Aladdin.CAPI.PKCS12.Java/src/aladdin/capi/pkcs12/PfxParentItem.java
package aladdin.capi.pkcs12;
import aladdin.*; 
import aladdin.capi.*;
import aladdin.asn1.iso.pkcs.pkcs12.*; 
import java.io.*; 
import java.util.*; 

///////////////////////////////////////////////////////////////////////////
// Неконечный элемент контейнера PKCS12 
///////////////////////////////////////////////////////////////////////////
public abstract class PfxParentItem extends PfxItem implements Iterable<PfxItem>
{
	// дочерние элементы
	protected List<PfxItem> items;		

	// конструктор
	protected PfxParentItem() { items = new ArrayList<PfxItem>(); }

    // освободить выделенные ресурсы
    @Override protected void onClose() throws IOException  
    {
        // освободить выделенные ресурсы
        for (PfxItem item : items) RefObject.release(item); super.onClose();
    }
	// получить элемент коллекции
	public PfxItem get(int i) { return items.get(i); }  
			
	// получить размер коллекции
	public int size() { return items.size(); } 

    // перечислитель объектов
    @Override public final Iterator<PfxItem> iterator() { return items.iterator(); }
    
	// признак наличия открытых данных
	@Override public boolean hasDecryptedItems()
	{ 
		// для всех дочерних элементов
		for (PfxItem item : items) 
		{ 
			// проверить наличие шифрования
			if (item.hasDecryptedItems()) return true; 
		}
		return false; 
	}
    // признак наличия закрытых данных
	@Override public boolean hasEncryptedItems()
	{ 
		// для всех дочерних элементов
		for (PfxItem item : items) 
		{ 
			// проверить наличие шифрования
			if (item.hasEncryptedItems()) return true; 
		}
		return false; 
	}
	// расшифровать элемент
	@Override protected void decrypt(PfxDecryptor decryptor) throws IOException
	{
		// расшифровать все дочерние элементы
		for (PfxItem item : items) item.decrypt(decryptor); 
	}
    // изменить значение
    @Override protected void change() throws IOException
    {
        // изменить все дочерние элементы
        for (PfxItem item : items) item.change(); 
    }
	// изменение дочерних элементов
	protected void onItemsChange() throws IOException {}

	// найти требуемый элемент
	public PfxContainerSafeBag[] findObjects(PfxFilter callback)
	{
		// создать список найденных элементов
		List<PfxContainerSafeBag> objs = new ArrayList<PfxContainerSafeBag>(); 

		// для каждого элемента коллекции
		for (PfxItem item : this)
		{
			// для внутренней коллекции
			if (item instanceof PfxParentItem) 
			{
				// найти элементы внутренней коллекции
                PfxContainerSafeBag[] objects = ((PfxParentItem)item).findObjects(callback); 
                
                // добавить элементы внутренней коллекции
				objs.addAll(Arrays.asList(objects));
			}
			else if (item instanceof PfxSafeBag)
			{
				// извлечь значение элемента
				SafeBag safeBag = ((PfxSafeBag)item).decoded(); 

				// указать зашифрованное значение элемента
				if (safeBag == null) safeBag = (SafeBag)((PfxSafeBag)item).encoded(); 
                    
                // определить идентификатор элемента
                byte[] keyID = safeBag.localKeyID(); 

                // при отсутствии идентификатора
                if (keyID == null && safeBag.bagAttributes() != null)
                {
                    // получить закодированное представление атрибутов
                    byte[] encoded = safeBag.bagAttributes().encoded(); 

                    // создать алгоритм хэширования
                    try (Hash hash = new aladdin.capi.ansi.hash.SHA1())
                    {
                        // вычислить хэш-значение от атрибутов
                        keyID = hash.hashData(encoded, 0, encoded.length); 
                    }
                    // обработать неожидаемую ошибку
                    catch (IOException e) { throw new RuntimeException(e); }
                }
				try { 
					// проверить критерий поиска
					if (callback != null && !callback.isMatch(safeBag, keyID)) continue; 

                    // добавить элемент в список
                    objs.add(new PfxContainerSafeBag((PfxSafeBag)item, keyID)); 
                }
				catch (Throwable e) {}
			}
		}
        // вернуть найденные элементы
		return objs.toArray(new PfxContainerSafeBag[0]);
	}
	// добавить дочерние элементы
	void addObjects(List<PfxData<SafeBag>> safeBags, PfxEncryptor encryptor) 
        throws IOException
	{
		// создать новый элемент
		PfxSafeContents item = new PfxSafeContents(this, safeBags, encryptor); 

		// добавить новый элемент
		items.add(0, item); onItemsChange(); 
	}
	// добавить дочерний элемент
	void addObject(PfxData<SafeBag> data) throws IOException
	{
		// в зависимости от типа внутреннего объекта
		if (data.content.bagId().value().equals(aladdin.asn1.iso.pkcs.pkcs12.OID.BT_SAFE_CONTENTS))
		{
			// добавить объект в список
			items.add(0, new PfxContentsBag(this, data)); 
		}
		// добавить объект в список
		else items.add(0, new PfxSafeBag(this, data)); onItemsChange();
	}
	// удалить требуемый элемент
	public void removeObject(PfxSafeBag pfxSafeBag) throws IOException
	{
		// удалить требуемый элемент
		items.remove(pfxSafeBag); onItemsChange();
	}
}

File: Source/Aladdin.CAPI.PKCS12/CAPI/PKCS12/PfxParentItem.cs
using System.Collections;
using System.Collections.Generic;

namespace Aladdin.CAPI.PKCS12
{
	///////////////////////////////////////////////////////////////////////////
	// Неконечный элемент контейнера PKCS12 
	///////////////////////////////////////////////////////////////////////////
	public abstract class PfxParentItem : PfxItem, IEnumerable<PfxItem>
	{
		// дочерние элементы
		protected List<PfxItem> items;		

		// конструктор
		protected PfxParentItem() { items = new List<PfxItem>(); }

        // освободить выделенные ресурсы
        protected override void OnDispose()
        {
            // освободить выделенные ресурсы
            foreach (PfxItem item in items) RefObject.Release(item); 

            // вызвать базовую функцию
            base.OnDispose();
        }
		// получить элемент коллекции
		public PfxItem this[int i] { get { return items[i]; } }  
			
		// получить размер коллекции
		public int Length { get { return items.Count; } } 

		// перечислитель элементов
		public IEnumerator<PfxItem> GetEnumerator() { return items.GetEnumerator(); }

		// перечислитель элементов
		IEnumerator IEnumerable.GetEnumerator() { return items.GetEnumerator(); }

		// признак наличия открытых данных
		public override bool HasDecryptedItems { get 
		{ 
			// для всех дочерних элементов
			foreach (PfxItem item in items) 
			{ 
				// проверить наличие шифрования
				if (item.HasDecryptedItems) return true; 
			}
			return false; 
		}}
		// признак наличия закрытых данных
		public override bool HasEncryptedItems { get 
		{ 
			// для всех дочерних элементов
			foreach (PfxItem item in items) 
			{ 
				// проверить наличие шифрования
				if (item.HasEncryptedItems) return true; 
			}
			return false; 
		}}
		// расшифровать элемент
		protected internal override void Decrypt(PfxDecryptor decryptor)
		{
			// расшифровать все дочерние элементы
            foreach (PfxItem item in items) item.Decrypt(decryptor); 
		}
        // изменить значение
        protected internal override void Change()
        {
			// изменить все дочерние элементы
            foreach (PfxItem item in items) item.Change(); 
        }
		// изменение дочерних элементов
		protected internal virtual void OnItemsChange() {}

		// найти требуемый элемент
		public PfxContainerSafeBag[] FindObjects(PfxFilter callback)
		{
			// создать список найденных элементов
			List<PfxContainerSafeBag> objs = new List<PfxContainerSafeBag>(); 

			// для каждого элемента коллекции
			foreach (PfxItem item in this)
			{
				// для внутренней коллекции
				if (item is PfxParentItem) 
				{
					// найти элементы внутренней коллекции
					objs.AddRange(((PfxParentItem)item).FindObjects(callback));
				}
				else if (item is PfxSafeBag)
				{
					// извлечь расшифрованное значение элемента
					ASN1.ISO.PKCS.PKCS12.SafeBag safeBag = ((PfxSafeBag)item).Decoded; 

					// указать зашифрованное значение элемента
					if (safeBag == null) safeBag = (ASN1.ISO.PKCS.PKCS12.SafeBag)((PfxSafeBag)item).Encoded; 
                    
                    // определить идентификатор элемента
                    byte[] keyID = safeBag.LocalKeyID; 

                    // при отсутствии идентификатора
                    if (keyID == null && safeBag.BagAttributes != null) 
                    {
                        // получить закодированное представление атрибутов
                        byte[] encoded = safeBag.BagAttributes.Encoded; 

                        // создать алгоритм хэширования
                        using (Hash hash = new ANSI.Hash.SHA1())
                        {
                            // вычислить хэш-значение от атрибутов
                            keyID = hash.HashData(encoded, 0, encoded.Length); 
                        }
                    }
					try { 
						// проверить критерий поиска
						if (callback != null && !callback.IsMatch(safeBag, keyID)) continue; 

                        // добавить элемент в список
                        objs.Add(new PfxContainerSafeBag((PfxSafeBag)item, keyID)); 
					}
					catch {}
				}
			}
			return objs.ToArray();
		}
		// добавить дочерние элементы
		public void AddObjects(PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag>[] safeBags, PfxEncryptor encryptor)
		{
			// создать новый элемент
			PfxSafeContents item = new PfxSafeContents(this, safeBags, encryptor); 

			// добавить новый элемент
            items.Insert(items.Count, item); OnItemsChange(); 
		}
		// добавить дочерний элемент
		public void AddObject(PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag> safeBag)
		{
			// в зависимости от типа внутреннего объекта
			if (safeBag.Content.BagId.Value == ASN1.ISO.PKCS.PKCS12.OID.bt_safeContents)
			{
				// добавить объект в список
				items.Insert(items.Count, new PfxContentsBag(this, safeBag)); 
			}
			// добавить объект в список
            else items.Insert(items.Count, new PfxSafeBag(this, safeBag)); OnItemsChange();
		}
		// удалить требуемый элемент
		public void RemoveObject(PfxSafeBag pfxSafeBag)
		{
			// удалить требуемый элемент
            items.Remove(pfxSafeBag); OnItemsChange();
		}
	}
}
